internal/bufferedread/buffered_reader.go

```diff
@@ -103,6 +106,191 @@ func NewBufferedReader(object *gcs.MinObject, bucket gcs.Bucket, config *Buffere
+		entry := p.blockQueue.Peek()
+		block := entry.block
+		blockStart := block.AbsStartOff()
+		blockEnd := blockStart + block.Cap()
+
+		if offset < blockStart || offset >= blockEnd {
+			// Offset is either before or beyond this block – discard.
+			p.blockQueue.Pop()
+			entry.cancel()
+
+			if _, waitErr := block.AwaitReady(context.Background()); waitErr != nil && !errors.Is(waitErr, context.Canceled) {
+				logger.Warnf("prepareQueueForOffset: AwaitReady error during discard (offset=%d): %v", offset, waitErr)
+			}
+
+			p.blockPool.Release(block)
+		} else {
+			break
+		}
+	}
+}
+
+// ReadAt reads data from the GCS object into the provided buffer starting at
+// the given offset. It implements the gcsx.Reader interface.
+//
+// The read is satisfied by reading from in-memory blocks that are prefetched
+// in the background. The core logic is as follows:
+//  1. Detect if the read pattern is random. If so, and if the random read
+//     threshold is exceeded, it returns a FallbackToAnotherReader error.
+//  2. Prepare the internal block queue by discarding any stale blocks from the
+//     head of the queue that are before the requested offset.
+//  3. If the queue becomes empty (e.g., on a fresh read or a large seek), it
+//     initiates a "fresh start" to prefetch blocks starting from the current
+//     offset.
+//  4. It then enters a loop to fill the destination buffer:
+//     a. It waits for the block at the head of the queue to be downloaded.
+//     b. If the download failed or was cancelled, it returns an appropriate error.
+//     c. If successful, it copies data from the downloaded block into the buffer.
+//     d. If a block is fully consumed, it is removed from the queue, and a new
+//     prefetch operation is triggered to keep the pipeline full.
+//  5. The loop continues until the buffer is full, the end of the file is
+//     reached, or an error occurs.
+func (p *BufferedReader) ReadAt(ctx context.Context, inputBuf []byte, off int64) (gcsx.ReaderResponse, error) {
+	resp := gcsx.ReaderResponse{DataBuf: inputBuf}
+	reqID := uuid.New()
+	start := time.Now()
+	initOff := off
+	blockIdx := initOff / p.config.PrefetchBlockSizeBytes
+
+	var bytesRead int
+	var err error
+
+	logger.Tracef("%.13v <- ReadAt(offset=%d, len=%d, blockIdx=%d)", reqID, off, len(inputBuf), blockIdx)
+
+	if off >= int64(p.object.Size) {
+		err = io.EOF
+		return resp, err
+	}
+	if len(inputBuf) == 0 {
+		return resp, nil
+	}
+
+	defer func() {
+		dur := time.Since(start)
+		if err != nil && err != io.EOF {
+			logger.Errorf("%.13v -> ReadAt failed (offset=%d, len=%d, blockIdx=%d): err: %v (%v)", reqID, initOff, len(inputBuf), blockIdx, err, dur)
+		} else {
+			logger.Tracef("%.13v -> ReadAt OK (offset=%d, len=%d, read=%d) (%v)", reqID, initOff, len(inputBuf), bytesRead, dur)
+		}
+	}()
+
+	if err = p.handleRandomRead(off); err != nil {
+		err = fmt.Errorf("ReadAt: handleRandomRead failed: %w", err)
+		return resp, err
+	}
+
+	prefetchTriggered := false
```

Comment:
Just thinking out loud here. Currently, we will be triggering only once for a given ReadAt call, this assumes size of ReadAt() < block-size.

What if size of ReadAt() > block-size: in this case one ReadAt() will basically evict many small blocks? Although this configuration is not recommended, but good to handle if we can. Ideally, we should trigger prefetch for every block eviction. Or may be we can ensure block-size can't be less than 1 MiB.